File: leap_year_tool.py
def is_leap_year(year: int) -> bool:
    if year % 400 == 0:
        return True
    if year % 100 == 0:
        return False
    return year % 4 == 0


def read_int(prompt: str) -> int | None:
    try:
        return int(input(prompt).strip())
    except ValueError:
        print("Invalid input. Please enter a valid integer year.")
        return None


def check_single_year() -> None:
    year = read_int("\nEnter a year: ")
    if year is None:
        return

    if is_leap_year(year):
        print(f"{year} is a leap year.")
    else:
        print(f"{year} is not a leap year.")


def check_year_range() -> None:
    start = read_int("\nEnter start year: ")
    if start is None:
        return

    end = read_int("Enter end year: ")
    if end is None:
        return

    if start > end:
        print("Start year cannot be greater than end year.")
        return

    leap_years = [str(y) for y in range(start, end + 1) if is_leap_year(y)]
    print(f"\nLeap years from {start} to {end}:")
    print(" ".join(leap_years) if leap_years else "None")


def main() -> None:
    while True:
        print("\n=== Leap Year Tool (Python) ===")
        print("1. Check a single year")
        print("2. List leap years in a range")
        print("3. Exit")
        choice = input("Choose an option (1-3): ").strip()

        if choice == "1":
            check_single_year()
        elif choice == "2":
            check_year_range()
        elif choice == "3":
            print("Goodbye!")
            break
        else:
            print("Please choose a valid option (1-3).")


if __name__ == "__main__":
    main()

File: leap_year_tool.c
#include <stdio.h>

int is_leap_year(int year) {
    if (year % 400 == 0) {
        return 1;
    }
    if (year % 100 == 0) {
        return 0;
    }
    return year % 4 == 0;
}

void check_single_year(void) {
    int year;
    printf("\nEnter a year: ");

    if (scanf("%d", &year) != 1) {
        printf("Invalid input. Please enter a valid integer year.\n");
        return;
    }

    if (is_leap_year(year)) {
        printf("%d is a leap year.\n", year);
    } else {
        printf("%d is not a leap year.\n", year);
    }
}

void check_year_range(void) {
    int start;
    int end;
    int i;
    int found = 0;

    printf("\nEnter start year: ");
    if (scanf("%d", &start) != 1) {
        printf("Invalid input for start year.\n");
        return;
    }

    printf("Enter end year: ");
    if (scanf("%d", &end) != 1) {
        printf("Invalid input for end year.\n");
        return;
    }

    if (start > end) {
        printf("Start year cannot be greater than end year.\n");
        return;
    }

    printf("\nLeap years from %d to %d:\n", start, end);
    for (i = start; i <= end; i++) {
        if (is_leap_year(i)) {
            printf("%d ", i);
            found = 1;
        }
    }

    if (!found) {
        printf("None");
    }
    printf("\n");
}

int main(void) {
    int choice;

    while (1) {
        printf("\n=== Leap Year Tool ===\n");
        printf("1. Check a single year\n");
        printf("2. List leap years in a range\n");
        printf("3. Exit\n");
        printf("Choose an option (1-3): ");

        if (scanf("%d", &choice) != 1) {
            printf("Invalid input. Exiting.\n");
            return 1;
        }

        switch (choice) {
            case 1:
                check_single_year();
                break;
            case 2:
                check_year_range();
                break;
            case 3:
                printf("Goodbye!\n");
                return 0;
            default:
                printf("Please choose a valid option (1-3).\n");
        }
    }
}
